fms/utils/generation.py

```diff
@@ -20,9 +20,73 @@ def _make_cache_contiguous(past_key_value_states):
     return n_kv_s
 
 
+def left_pad(input_ids: torch.Tensor, max_length: int, pad_id: int = 0) -> torch.Tensor:
+    """
+    left pad an input_ids tensor
+
+    Parameters
+    ----------
+    input_ids: torch.Tensor
+        input ids corresponding to a single sequence in a batch
+    max_length: int
+        the max length to pad to
+    pad_id: int
+        the token to set as a pad in the resulting tensor
+
+    Returns
+    -------
+    torch.Tensor
+        a left padded tensor
+    """
+    pads_tensor = torch.tensor(
+        [pad_id] * (max_length - input_ids.size(0)),
+        device=input_ids.device,
+        dtype=torch.long,
+    )
+    return torch.cat((pads_tensor, input_ids))
+
+
+def __create_attention_mask(
+    input_ids: torch.Tensor, use_cache: bool, is_prompt: bool, pad_id: int = -1
+) -> Optional[torch.Tensor]:
+    """
+    Optionally create an attention mask if one is required
+
+    Parameters
+    ----------
+    input_ids: torch.Tensor
+        the input ids to the model's forward method
+    use_cache: bool
+        if True caching is being used, otherwise caching is not being used
+    is_prompt: bool
+        if True, input_ids correspond to the prompt, otherwise input_ids contain generated tokens
+    pad_id: int
+        the pad id used when padding. If the pad_id is -1, no padding is being used, otherwise assume some padding
+
+    Returns
+    -------
+    Optional[torch.Tensor]
+        if is_prompt, no use_cache, batch required padding => a bool Tensor is created for the mask
+        otherwise None is returns and the mask will be handled internally as part of sdpa
+    """
+    # compute the attention mask
+    # always need a mask if this is the prompt or if not using a cache
+    if is_prompt or not use_cache:
+        is_pad = input_ids == pad_id
```

Comment:
I'm still not convinced this is a valid way to construct the mask. the token being used for padding may also be a valid token within the prompt. llama doesn't have an explicit pad token, so a user might decide to pad with whitespace or bos or eos or unk, any of which could be part of valid model input/output.

Reply:
One option is we just construct this based on the sizes of the input tensors in the list. When a user provides a list, we are considering this a batch.

Reply:
yeah that was why I had previously suggested rearranging to pass the whole list of tensors. passing the pre-pad length of each row in the batch might be a bit awkward, but maybe there's some other way to do it.